File: dotnet/usb-client-python/serial_protocol_handler.py
import json
import threading

import serial

from eng_board_resources import (
    EnginePower,
    SystemPower,
    TransformerPower,
)
from utils import log_debug


class SerialProtocolHandler:
    """
    A class to handle serial communication with a specific protocol.
    Methods:
        is_connected() -> bool:
            Checks if the serial port is open and the connection is valid.
        reset_port():
            Resets the input and output buffers of the serial port.
        establish_connection() -> bool:
            Establishes a connection by sending an initialization command and checking the response.
        query_engine_power() -> list[EnginePower]:
            Queries the engine power data from the device.
        query_transformer_power() -> list[TransformerPower]:
            Queries the transformer power data from the device.
        query_system_power() -> list[SystemPower]:
            Queries the system power data from the device.
        query_list(protocol_token: str) -> list:
            Sends a query command and returns the parsed response as a list.
        set_engine_power(data: list[EnginePower]) -> str:
            Sends a set command with engine power data.
        set_transformer_power(data: list[TransformerPower]) -> str:
            Sends a set command with transformer power data.
        set_system_power(data: list[SystemPower]) -> str:
            Sends a set command with system power data.
        set_list(protocol_token: str, data: list) -> str:
            Sends a set command with the given data and returns the response.
    """

    DEFAULT_BAUD_RATE = 9600

    SER_PROTO_RESPONSE_DELIMITER = "|"
    SER_PROTO_QUERY = "_QUERY"
    SER_PROTO_RESPONSE = "_RESPONSE"
    SER_PROTO_SET = "_SET"

    SER_PROTO_INIT_HEADER = "SP_INIT"
    SER_PROTO_INIT_RESPONSE = "SP_READY"
    SER_PROTO_OK = "SP_OK"
    SER_PROTO_ERR = "SP_ERR;"
    SER_PROTO_ENGINE_POWER = "SP_ENG_POWER"
    SER_PROTO_TRANSFORMER_POWER = "SP_TRANS_POWER"
    SER_PROTO_SYSTEM_POWER = "SP_SYS_POWER"

    def __init__(self, port_name: str, board_name: str):
        self.port_name = port_name
        self.board_name = board_name
        self.communication_lock = threading.Lock()
        self.is_valid = False

        self.port = serial.Serial(
            port=port_name,
            baudrate=self.DEFAULT_BAUD_RATE,
            parity=serial.PARITY_NONE,
            bytesize=serial.EIGHTBITS,
            stopbits=serial.STOPBITS_ONE,
            timeout=0.5,
            write_timeout=0.5,
        )
        self.port.dtr = True

    @property
    def is_connected(self) -> bool:
        return self.port.is_open and self.is_valid

    def reset_port(self):
        self.port.reset_input_buffer()
        self.port.reset_output_buffer()

    def establish_connection(self) -> bool:
        if self.port is None:
            return False

        with self.communication_lock:
            try:
                self.reset_port()
                self.port.write((self.SER_PROTO_INIT_HEADER + "\n").encode("utf-8"))
                response = self.port.readline().decode("utf-8").strip()

                if response != self.SER_PROTO_INIT_RESPONSE:
                    print(
                        f"ERR: {self} is not a SerProto client: Invalid init response: {response}"
                    )
                    return False

                self.is_valid = True
                return True
            except serial.SerialTimeoutException:
                print(
                    f"ERR: {self} is not a SerProto client: Timeout during establish_connection"
                )
            except Exception as ex:
                print(
                    f"ERR: {self} is not a SerProto client: Unknown exception in establish_connection: {str(ex)}",
                    exc_info=True,
                )

            return False

    def query_engine_power(self) -> list[EnginePower]:
        log_debug("Sending QueryEnginePower")
        return [
            EnginePower.from_json(jsonBlob)
            for jsonBlob in self.query_list(self.SER_PROTO_ENGINE_POWER)
        ]

    def query_transformer_power(self) -> list[TransformerPower]:
        log_debug("Sending QueryTransformerPower")
        return [
            TransformerPower.from_json(jsonBlob)
            for jsonBlob in self.query_list(self.SER_PROTO_TRANSFORMER_POWER)
        ]

    def query_system_power(self) -> list[SystemPower]:
        log_debug("Sending QuerySystemPower")
        return [
            SystemPower.from_json(jsonBlob)
            for jsonBlob in self.query_list(self.SER_PROTO_SYSTEM_POWER)
        ]

    def query_list(self, protocol_token: str) -> list:
        empty_result = []

        with self.communication_lock:
            try:
                self.reset_port()
                protocol_command = protocol_token + self.SER_PROTO_QUERY

                log_debug(f"DBG: QueryList: Sending: {protocol_command}")
                self.port.write((protocol_command + "\n").encode("utf-8"))
                response = self.port.readline().decode("utf-8").strip()
                log_debug(f"DBG: QueryList: Received: {response}")

                response_parts = response.split(self.SER_PROTO_RESPONSE_DELIMITER)

                if (
                    len(response_parts) != 2
                    or response_parts[0] != protocol_token + self.SER_PROTO_RESPONSE
                ):
                    print(
                        f"ERR: SerProto {self.port_name}({self.board_name}): Invalid query list response: {response}"
                    )
                    return empty_result

                try:
                    return json.loads(response_parts[1])
                except Exception as ex:
                    print(
                        f"ERR: SerProto {self.port_name}({self.board_name}): Invalid query list response format: {response}: {str(ex)}",
                        exc_info=True,
                    )
                    return empty_result
            except serial.SerialTimeoutException:
                print(
                    f"ERR: SerProto {self.port_name}({self.board_name}): Timeout during query list"
                )

            return empty_result

    def set_engine_power(self, data: list[EnginePower]) -> str:
        json_blob = [EnginePower.to_json(enginePower) for enginePower in data]
        return self.set_list(self.SER_PROTO_ENGINE_POWER, json_blob)

    def set_transformer_power(self, data: list[TransformerPower]) -> str:
        json_blob = [
            TransformerPower.to_json(transformerPower) for transformerPower in data
        ]
        return self.set_list(self.SER_PROTO_TRANSFORMER_POWER, json_blob)

    def set_system_power(self, data: list[SystemPower]) -> str:
        json_blob = [SystemPower.to_json(systemPower) for systemPower in data]
        return self.set_list(self.SER_PROTO_SYSTEM_POWER, json_blob)

    def set_list(self, protocol_token: str, data: list) -> str:
        with self.communication_lock:
            try:
                self.reset_port()

                request = (
                    protocol_token
                    + self.SER_PROTO_SET
                    + self.SER_PROTO_RESPONSE_DELIMITER
                    + json.dumps(data)
                )
                self.port.write((request + "\n").encode("utf-8"))
                response = self.port.readline().decode("utf-8").strip()

                if response != self.SER_PROTO_OK and not response.startswith(
                    self.SER_PROTO_ERR
                ):
                    print(f"ERR: {self}: Invalid set list response: {response}")
                    return self.SER_PROTO_ERR

                return response
            except serial.SerialTimeoutException:
                print(f"ERR: {self}: Timeout during set list")
                return self.SER_PROTO_ERR + ":Timeout"

    def __str__(self):
        return f"SerProto:{self.port_name}({self.board_name})"

    def __del__(self):
        self.dispose()

    def dispose(self):
        log_debug(f"{self}: Disposing")
        self.is_valid = False
        if self.port.is_open:
            log_debug(f"{self}: Closing SerialPort")
            self.port.close()
        log_debug(f"{self}: Disposing SerialPort")
        del self.port

File: engineering-board-usb-client-csharp/CircuitPythonBoardManager/SerialProtocolHandler.cs
using System.IO.Ports;
using System.Text;
using EngBoard;
using Helpers;

namespace CircuitPythonInterface;

public class SerialProtocolHandler : IDisposable
{
    public static int DEFAULT_BAUD_RATE { get; set; } = 9600;

    public const string SER_PROTO_RESPONSE_DELIMITER = "|";
    public const string SER_PROTO_QUERY = "_QUERY";
    public const string SER_PROTO_RESPONSE = "_RESPONSE";
    public const string SER_PROTO_SET = "_SET";

    public const string SER_PROTO_INIT_HEADER = "SP_INIT";
    public const string SER_PROTO_INIT_RESPONSE = "SP_READY";
    public const string SER_PROTO_OK = "SP_OK";
    public const string SER_PROTO_ERR = "SP_ERR;";
    public const string SER_PROTO_ENGINE_POWER = "SP_ENG_POWER";
    public const string SER_PROTO_TRANSFORMER_POWER = "SP_TRANS_POWER";
    public const string SER_PROTO_SYSTEM_POWER = "SP_SYS_POWER";

    private readonly object CommunicationLock = new();

    public readonly string PortName;
    public readonly string BoardName;
    private readonly SerialPort port;
    private bool IsValid = false;

    public SerialProtocolHandler(string portName, string boardName)
    {
        PortName = portName;
        BoardName = boardName;

        port = new SerialPort(portName, DEFAULT_BAUD_RATE, Parity.None, 8, StopBits.One)
        {
            ReadTimeout = 500,
            WriteTimeout = 500,
            Encoding = Encoding.UTF8,
            DtrEnable = true,
        };

        port.Open();
    }

    public bool IsConnected => port.IsOpen && IsValid;

    private void ResetPort()
    {
        // Make sure there is no pending data from other (possibly failed) transactions
        port.DiscardInBuffer();
        port.DiscardOutBuffer();
    }

    public bool EstablishConnection()
    {
        if (port == null) { return false; }

        lock (CommunicationLock)
        {
            try
            {
                ResetPort();

                port.WriteLine(SER_PROTO_INIT_HEADER);
                string response = port.ReadLine();

                if (response != SER_PROTO_INIT_RESPONSE)
                {
                    Log.Error($"{this} is not a SerProto client: Invalid init response: {response}");
                    return false;
                }

                IsValid = true;
                return true;
            }
            catch (TimeoutException)
            {
                Log.Error($"{this} is not a SerProto client: Timeout during EstablishConnection");
            }
            catch (Exception ex)
            {
                Log.Error(ex, $"{this} is not a SerProto client: Unknown exception EstablishConnection");
            }

            return false;
        }
    }

    public List<EnginePower> QueryEnginePower()
    {
        Log.Information($"Sending QueryEnginePower");
        return QueryList<EnginePower>(SER_PROTO_ENGINE_POWER);
    }

    public List<TransformerPower> QueryTransformerPower()
    {
        Log.Information($"Sending QueryTransformerPower");
        return QueryList<TransformerPower>(SER_PROTO_TRANSFORMER_POWER);
    }

    public List<SystemPower> QuerySystemPower()
    {
        Log.Information($"Sending QuerySystemPower");
        return QueryList<SystemPower>(SER_PROTO_SYSTEM_POWER);
    }

    private List<T> QueryList<T>(string protocolToken) where T : EngBoardResource
    {
        List<T> EmptyResult = [];

        lock (CommunicationLock)
        {
            try
            {
                ResetPort();
                string protocolCommand = protocolToken + SER_PROTO_QUERY;

                Log.Debug($"QueryList: Sending: {protocolCommand}");
                port.WriteLine(protocolCommand);
                string response = port.ReadLine();
                Log.Debug($"QueryList: Received: {response}");

                string[] responseParts = response.Split(SER_PROTO_RESPONSE_DELIMITER);

                if (responseParts.Length != 2 || responseParts[0] != protocolToken + SER_PROTO_RESPONSE)
                {
                    Log.Error($"SerProto {PortName}({BoardName}): Invalid query list response: {response}");
                    return EmptyResult;
                }

                try {
                    return JsonConvert.DeserializeObject<List<T>>(responseParts[1]) ?? EmptyResult;
                } catch (Exception ex) {
                    Log.Error(ex, $"SerProto {PortName}({BoardName}): Invalid query list response format: {response}: {ex.Message}");
                    return EmptyResult;
                }   
            }
            catch (TimeoutException)
            {
                Log.Error($"SerProto {PortName}({BoardName}): Timeout during query list");
            }

            return EmptyResult;
        }
    }

    public string SetEnginePower(List<EnginePower> data)
    {
        return SetList(SER_PROTO_ENGINE_POWER, data);
    }

    public string SetTransformerPower(List<TransformerPower> data)
    {
        return SetList(SER_PROTO_TRANSFORMER_POWER, data);
    }

    public string SetSystemPower(List<SystemPower> data)
    {
        return SetList(SER_PROTO_SYSTEM_POWER, data);
    }

    public string SetList<T>(string protocolToken, List<T> data) where T : EngBoardResource
    {
        lock (CommunicationLock)
        {
            try
            {
                ResetPort();

                string request = protocolToken + SER_PROTO_SET + SER_PROTO_RESPONSE_DELIMITER
                    + JsonConvert.SerializeObject(data);
                port.WriteLine(request);
                string response = port.ReadLine();

                if (response != SER_PROTO_OK && !response.StartsWith(SER_PROTO_ERR))
                {
                    Log.Error($"{this}: Invalid set list response: {response}");
                    return SER_PROTO_ERR;
                }

                return response;
            }
            catch (TimeoutException)
            {
                Log.Error($"{this}: Timeout during set list");
                return SER_PROTO_ERR + ":Timeout";
            }
        }
    }

    public override string ToString()
    {
        return $"SerProto:{PortName}({BoardName})";
    }

    private bool AlreadyDisposed;

    protected virtual void Dispose(bool disposing)
    {
        Log.Information($"{this}: Disposing");
        if (!AlreadyDisposed)
        {
            if (disposing)
            {
                IsValid = false;
                if (port.IsOpen)
                {
                    Log.Information($"{this}: Closing SerialPort");
                    port.Close();
                }
                Log.Information($"{this}: Disposing SerialPort");
                port.Dispose();
            }

            AlreadyDisposed = true;
        }
    }

    public void Dispose()
    {
        // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
        Dispose(disposing: true);
        GC.SuppressFinalize(this);
    }
}